libr/util/buf.c

```diff
@@ -555,31 +558,40 @@ R_API st64 r_buf_fwrite(RBuffer *b, const ut8 *buf, const char *fmt, int n) {
 
 R_API st64 r_buf_fwrite_at(RBuffer *b, ut64 addr, const ut8 *buf, const char *fmt, int n) {
 	r_return_val_if_fail (b && buf && fmt && !b->readonly, -1);
+	st64 o_addr = r_buf_seek (b, 0, R_BUF_CUR);
 	st64 r = r_buf_seek (b, addr, R_BUF_SET);
 	if (r < 0) {
 		return r;
 	}
-	return r_buf_fwrite (b, buf, fmt, n);
+	r = r_buf_fwrite (b, buf, fmt, n);
+	r_buf_seek (b, o_addr, R_BUF_SET);
+	return r;
 }
 
 R_API st64 r_buf_read_at(RBuffer *b, ut64 addr, ut8 *buf, ut64 len) {
 	r_return_val_if_fail (b && buf, -1);
+	st64 o_addr = r_buf_seek (b, 0, R_BUF_CUR);
 	st64 r = r_buf_seek (b, addr, R_BUF_SET);
 	if (r < 0) {
 		return r;
 	}
 
-	return r_buf_read (b, buf, len);
+	r = r_buf_read (b, buf, len);
+	r_buf_seek (b, o_addr, R_BUF_SET);
+	return r;
 }
 
 R_API st64 r_buf_write_at(RBuffer *b, ut64 addr, const ut8 *buf, ut64 len) {
 	r_return_val_if_fail (b && buf && !b->readonly, -1);
+	st64 o_addr = r_buf_seek (b, 0, R_BUF_CUR);
 	st64 r = r_buf_seek (b, addr, R_BUF_SET);
 	if (r < 0) {
 		return r;
 	}
 
-	return r_buf_write (b, buf, len);
+	r = r_buf_write (b, buf, len);
```

Comment:
isnt r_buf_write suposed to increase the seek ? why do we have to call rbufseek in so many places?

Reply:
yes, `r_buf_write` increase the seek, but the following `r_buf_seek` restores the original seek to the place it was before doing the `r_buf_write_at`. Suppose you are doing:
```C
r_buf_seek(b, 0, R_BUF_SET);
r_buf_read(b, buf, 10);
r_buf_read(b, buf2, 1);
r_buf_read(b, buf3, 5);
```

then you need to read at some random place, so you would need to do `r_buf_read_at`, but without this patch `r_buf_read_at` would change the seek, so you can't go back to reading sequentially. With this patch, instead, you are sure that after doing a `r_buf_read_at`/`r_buf_write_at` the seek is not changed.